### ISSUE TITLE
dfu-programmer: no device present.

### ISSUE THREAD

**Issue by genokolar** [qmk/qmk_toolbox#58]
the QMK Toolbox can find DFU device connected，but when press Flash, the dfu-programmer can't find the device.
if  build the firmware use this command: make keyboard:default:dfu.  no problem
------------
*** QMK Toolbox (https://qmk.fm/toolbox)
    Supporting following bootloaders:
     - DFU (Atmel, LUFA) via dfu-programmer (http://dfu-programmer.github.io/)
     - Caterina (Arduino, Pro Micro) via avrdude (http://nongnu.org/avrdude/)
     - Halfkay (Teensy, Ergodox EZ) via teensy_loader_cli (https://pjrc.com/teensy/loader_cli.html)
     - STM32 (ARM) via dfu-util (http://dfu-util.sourceforge.net/)
     - Kiibohd (ARM) via dfu-util (http://dfu-util.sourceforge.net/)
     - BootloadHID (Atmel, ps2avrGB, CA66) via bootloadHID (https://www.obdev.at/products/vusb/bootloadhid.html)
     - Atmel SAM-BA via mdloader (https://github.com/patrickmt/mdloader)
    And the following ISP flasher protocols:
     - USBTiny (AVR Pocket)
     - AVRISP (Arduino ISP)
*** DFU device connected: ATm32U4DFU -- 03EB:2FF4:0001 {88bae032-5a81-49f0-bc3d-a4ff138216d6}
*** Attempting to flash, please don't remove device
>>> dfu-programmer.exe atmega32u4 erase --force
    dfu-programmer: no device present.
>>> dfu-programmer.exe atmega32u4 flash "C:\Users\geno\OneDrive\Keyboard\firmware\tu60rgb_geno.hex"
    dfu-programmer: no device present.
>>> dfu-programmer.exe atmega32u4 reset
    dfu-programmer: no device present.

**Comment by genokolar** [comment]
fixed it。only need replace the dfu-programmer。

https://github.com/genokolar/QMK_Keyboard/blob/master/qmk_toolbox.exe

**Comment by anjari** [comment]
I also met same issue. The connected device is using driver ATm32U4DFU, but QMK Toolbox expects driver ATmega32U4. This could happen if you're using TMK previously. Even if you had installed the ATmega32U4 driver, it won't replace the existing ATm32U4DFU driver. So the solution is uninstall the ATm32U4DFU driver.

**Comment by drashna** [comment]
This should be handled better, now, and may no longer be an issue.

**Comment by lacymorrow** [comment]
Still an issue :(2

**Comment by fauxpark** [comment]
@lacymorrow this page may help: https://docs.qmk.fm/#/driver_installation_zadig

**Comment by lacymorrow** [comment]
Thank you @fauxpark, I'll try tonight on my PC. (I was able to flash without issues on macOS).

**Comment by caksoylar** [comment]
I ran into the same issue today with an Iris rev. 4 board with ATmega32U4 controller. Installing the `libusb-win32` driver is installed with zadig resolved the issue. I also hadn't installed TMK or any previous version of QMK before.

**Comment by drashna** [comment]
There have been a number of changes in this regard recently, including prompting to install drivers on first run.  

This should be a lot better, now.

**Comment by amowry** [comment]
My customers are using QMK Toolbox to install firmware on a MIDI device that I make, and some are reporting this problem. I'm using a Caterina bootloader, so no driver should be necessary on Windows 10. Presumably a driver has been installed? If this is the problem, could someone recommend the easiest way to uninstall the driver?

**Comment by fauxpark** [comment]
@amowry if the Toolbox is running dfu-programmer, then it's definitely not Caterina, so the driver will need to be installed.

**Comment by amowry** [comment]
Sorry, I probably should have started a new issue instead of replying to this one-- my device is Caterina and so uses AVRDude instead of dfu-programmer, but my users are also getting the "no device present" warning in some cases. I assume it's because they somehow have had a driver installed.

**Comment by fauxpark** [comment]
That's not much to go on, on its own. Some screenshots and steps followed would be helpful.

It may be worth noting that the device needs to be reset into bootloader mode by shorting the RST and GND pins, or tapping the reset button (possibly twice in quick succession). It will stay in bootloader mode for about 8 seconds. Caterina devices should appear as a serial port, and do not need any drivers as they use the builtin `usbser` driver on Windows. On Linux, the ModemManager service will need to be disabled as it hogs all the serial ports it finds. With systemd: `sudo systemctl disable --now ModemManager.service`

**Comment by amowry** [comment]
Thank you-- I was going to have users send my screenshots of the error, but it seems to have resolved itself by uninstalling QMK Toolbox, restarting the machine, and then reinstalling QMK Toollbox. I suggested this to them because I actually had the same thing happen on one of my Windows machines, and this procedure solved it. So, maybe it wasn't a driver issue as I had thought. I'll report back, hopefully with more info,  if I hear of anyone else encountering this. Thanks again!!

**Comment by rendomnet** [comment]
same problem here

**Comment by pirj** [comment]
Prefixing with `sudo` helps.

**Comment by rogers140** [comment]
> I ran into the same issue today with an Iris rev. 4 board with ATmega32U4 controller. Installing the `libusb-win32` driver with zadig resolved the issue. I also hadn't installed TMK or any previous version of QMK before.

This fixed my issue with ATmega32U4  (Tokyo60). 

1. First you need to uninstall the driver from Device Manager, 
2. Unplug and replug, enter into bootloader mode (L-shift + right-shift + fn + b) 
3. Follow https://docs.qmk.fm/#/driver_installation_zadig?id=installation here to load libusb-win32 into the recognized device (the one probably named ATm32U4DFU)
4. Unplug and replug. 
5. Then go to QMK toolbox, enter bootloader mode, should see something like "Atmel DFU device connected (libusb0): Atmel Corp. ATm32U4DFU (03EB:2FF4:0000)"
6. Then you can flash the VIA hex.

**Comment by ashokkbhaskar** [comment]
I solved the `dfu-programmer: no device present` errors I was getting when running `qmk flash` by plugging my keyboard (gBoards Georgi) into my laptop directly instead of into my USB hub for the flash process.

**Comment by fauxpark** [comment]
I believe the driver situation is significantly improved now, so I'm closing this. If there are any further issues just comment here.

**Comment by danysko5** [comment]
> > I ran into the same issue today with an Iris rev. 4 board with ATmega32U4 controller. Installing the `libusb-win32` driver with zadig resolved the issue. I also hadn't installed TMK or any previous version of QMK before.
> 
> This fixed my issue with ATmega32U4 (Tokyo60).
> 
> 1. First you need to uninstall the driver from Device Manager,
> 2. Unplug and replug, enter into bootloader mode (L-shift + right-shift + fn + b)
> 3. Follow https://docs.qmk.fm/#/driver_installation_zadig?id=installation here to load libusb-win32 into the recognized device (the one probably named ATm32U4DFU)
> 4. Unplug and replug.
> 5. Then go to QMK toolbox, enter bootloader mode, should see something like "Atmel DFU device connected (libusb0): Atmel Corp. ATm32U4DFU (03EB:2FF4:0000)"
> 6. Then you can flash the VIA hex.

have the same issue rn, cant install driver with zadig. always get error

**Comment by therightnee** [comment]
Hello! I am running into an error that seems very similar to the issue described above. Working with a ATMega32U4 (Ploopy Thumb Trackball) and consistently getting this error via QMK Toolbox. 

`Atmel DFU device connected (libusb0): Atmel Corp. ATmega32U4 (03EB:2FF4:0000)
Attempting to flash, please don't remove device
> dfu-programmer.exe atmega32u2 erase --force
> dfu-programmer: no device present.
> dfu-programmer.exe atmega32u2 flash --force "C:\Users\theri\qmk_firmware\.build\ploopyco_trackball_thumb_rev1_001_default.hex"
> dfu-programmer: no device present.
> dfu-programmer.exe atmega32u2 reset
> dfu-programmer: no device present.
Flash complete`

As best as I can tell it is using the right driver, but thought I would still reach out and ask. Thanks in advance!

**Comment by sigprof** [comment]
@therightnee Looks like you have a wrong MCU selected in the dropdown — ATmega32U**2** instead of ATmega32U**4**.

**Comment by therightnee** [comment]
Ah apologies - I tried different selections with the same results, I just happened to copy that error message. Here is what I get with the atmega32u4 selection.

`Atmel DFU device connected (libusb0): Atmel Corp. ATmega32U4 (03EB:2FF4:0000)
Attempting to flash, please don't remove device
> dfu-programmer.exe atmega32u4 erase --force
> Erasing flash...  ERROR
> dfu-programmer.exe atmega32u4 flash --force "C:\Users\theri\qmk_firmware\.build\ploopyco_trackball_thumb_rev1_001_default.hex"
> dfu-programmer: no device present.
> dfu-programmer.exe atmega32u4 reset
> dfu-programmer: no device present.
Flash complete
Attempting to flash, please don't remove device
> dfu-programmer.exe atmega32u4 erase --force
> Erasing flash...  ERROR
> dfu-programmer.exe atmega32u4 flash --force "C:\Users\theri\qmk_firmware\.build\ploopyco_trackball_thumb_rev1_001_default.hex"
> dfu-programmer: no device present.
> dfu-programmer.exe atmega32u4 reset
> dfu-programmer: no device present.
Flash complete
`

**Comment by fauxpark** [comment]
Could you try running that first erase command in QMK MSYS with `--debug=999`?

**Comment by therightnee** [comment]
@fauxpark - Sorry, I am not sure what you are asking? 

I am happy to report that I got my device to flash, but I cannot explain how or why what I did works. When I received the errors posted above, I was using a USB-C to USB-C cable (not PD compatible). I switched to a USB-A 3.1 to USB-C cable, and tried the nearest port to the previous USB-C port I was using, but I ran into some bad luck there and I now believe that port is non-functional.

Keeping with the USB-A 3.1 to USB-C cable, I plugged it into a USB-A port on a hub inside my monitor, which connects to the laptop via a single USB-C cable. With this configuration (USB-C on the mouse, USB-A to the hub, USB-C back to the laptop), I was able to successfully flash my .hex file and the mouse is functional now. 

I am able to use the mouse with both USB-C to USB-C and USB-C to USB-A 3.1 cables.
